lib.strings.concatMapAttrsStringSep: init

ShamrockLee opened the pull request:
## Description of changes

This pull request introduces `lib.concatMapAttrsStringSep`, the attribute set version of `lib.concatMapAttrsStringSet`. This function is initially defined locally in the Nix expression of `apptainer` and `singularity` packages. It increases the readability of attribute-set-to-string expressions.

**Before**

```nix
lib.concatStringsSep "\n" (lib.attrValues (lib.mapAttrs (name: package: "${name}: ${package.version}") packages))
```

or

```nix
lib.concatMapStringsSep "\n" (name: "${name}: ${packages.${name}.version}") (lib.attrNames packages)
```

**After**
```nix
lib.concatMapAttrsStringSep "\n" (name: package: "${name}: ${package.version}") packages
```

Cc:
@infinisil for the new library function
@roberth for the restructuring of `tests.overriding`

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux (only the manuals are rebuilt)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
lib.strings.concatMapAttrsStringSep: init

ShamrockLee committed:
apptainer: singularity: use lib.concatMapAttrsStringSep

Use concatMapAttrsStringSep helper function from the standard library
instead of the locally-improvised one.

ShamrockLee commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

ShamrockLee commented:
> `concatMapStringsSep` has put us on a [slippery slope](https://wiki.haskell.org/Fairbairn_threshold)

@roberth Thank you for sharing insight into the structural problem of adding all these composite functions. Providing reusable and composable parts is indeed better than pouring all the combinations of `concat[I][Map]String[Attr]s[Sep]` into `lib.strings`.

> 4\. Wait for Nixpkgs to bump `lib/minver.nix`. There's 2.3 hold-outs in the community, so may have to compromise on a backport and point release.

I didn't find explicit documentation about the lifecycle of each Nix version and the backport policy, and Nixpkgs's `lib/minver.nix` seems to have stayed at `2.3` for years. (Is it a result of the 2.4 CLI changing controversy?) Does Nixpkgs adopt any new Language features developed after Nix 2.3?

> ```nix
> concatStringsSep "\n" <| unique <| mapAttrs (k: v: f k v) <| attrs
> ```

As for the current standard library, `mapAttrs` maps between attribute sets, and the equivalent representation of `concatMapStringAttrsSep (k: v: f k v) attrs` with `mapAttrs` would be

```nix
concatStringsSep "\n" (attrValues (mapAttrs (k: v: f k v) attrs))
```

With the pipe operator, it would be

```nix
concatStringsSep "\n" <| attrValues <| mapAttrs  (k: v: f k v) <| attrs
```

Or the other direction.

I'll restructure the `tests.overriding.tests` with the `mapAttrs`-based pattern in another PR so that the test writing experience of overriding-related features would be smoother.

ShamrockLee commented:
@fricklerhandwerk I just remembered that I need to add double stars at the beginning of the comment block (`/**`) to make the documentation comment in the new format build. Sorry for the noise.

ShamrockLee committed:
tests.overriding: simplify implementation with lib.concatMapStringAttrsSep

infinisil commented on lib/strings.nix:
````diff
@@ -269,6 +269,37 @@ rec {
+    # Type
+
+    ```
+    concatMapStringAttrsSep :: String -> (String: Any: String) -> AttrSet -> String
````
```suggestion
    concatMapStringAttrsSep :: String -> (String -> Any -> String) -> AttrSet -> String
```

infinisil commented on lib/strings.nix:
```diff
@@ -269,6 +269,37 @@ rec {
+
+    :::
+  */
+  concatMapStringAttrsSep =
```
Shouldn't this be

```suggestion
  concatMapAttrsStringSep =
```

?

ShamrockLee replied:
Cool!

infinisil reviewed (changes requested):
Needs a test, but other than that I'm totally in favor of having this!

@roberth Let's keep the model of diverging, then converging in mind. I think we are in the diverging phase with a lot of `lib`, where we don't have good answers to everything, so it's fine to experiment a bit, even in the stable namespace (but only with long, unambiguous identifiers!). Though having an [experimental `lib`](https://github.com/NixOS/rfcs/pull/82) would be great too for this, but that's a future topic :)

ShamrockLee commented:
> Needs a test, but other than that, I'm totally in favor of having this!

Glad you like it! There is a test inside [lib/tests/misc.nix](https://github.com/NixOS/nixpkgs/pull/330010/files#diff-9c1b3fea9dfcb1a6c1154253a234e794b1c2e49c4c84421bac388786ad0615de).

ShamrockLee commented:
I just noticed that `lib.mapAttrs` is implemented by a built-in function.

Should I reimplement `lib.concatMapAttrsStringSep` with `lib.mapAttrs`?

ShamrockLee commented:
> I renamed it `concatMapAttrsStringSep` as @infinisil suggested.
> 
> I noticed that `lib.mapAttrs` is implemented by a built-in function. Should I reimplement `lib.concatMapAttrsStringSep` with `lib.mapAttrs`? The "before" part of the issue description shows both candidate implementations.

The renaming somehow doesn't get pushed. I'll rebase the feature branch and push it up.

infinisil commented on lib/strings.nix:
```diff
@@ -269,6 +269,37 @@ rec {
+    :::
+  */
+  concatMapAttrsStringSep =
+    # Separator to add between items
```
Docs here don't get rendered anymore, check out other functions in this file to see how inputs are declared in the docs string

ShamrockLee replied:
Addressed.

infinisil reviewed (approved):
I think this is good other than the one comment I left!

ShamrockLee commented:
@infinisil Thank you for your approval.

I rebased the feature branch to resolve conflicts, adopted the `mapAttrs`-based implementation from @roberth, and adjusted some wording in the argument documentation.

The `mapAttrs`-based implementation is shorter and more readable. It also uses more `builtins`-backed functions, which would make it more efficient (I haven't benchmarked it yet).
